Add copy out to the fallback path in SR invocation of composed op

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70656
* #70464
* __->__ #70871
* #70508
* #70463

We had previously handled reusing memory in the optimized kernel execution path, but not yet handled it if we hit the unoptimized fallback.

Differential Revision: [D33458652](https://our.internmc.facebook.com/intern/diff/D33458652)

Commit:
Add copy out to the fallback path in SR invocation of composed op

[ghstack-poisoned]

navahgar commented on test/cpp/tensorexpr/test_kernel.cpp:
```diff
@@ -1551,6 +1552,46 @@ TEST_F(Kernel, RunWithAllocatedOutputs) {
+  ASSERT_TRUE(at::allclose(out, ref));
+
+  auto inp_1 = at::ones({4, 4}, TensorOptions(kCPU).dtype(at::kFloat));
+  auto inp_ref = at::ones({4, 4}, TensorOptions(kCPU).dtype(at::kFloat));
```
unused

navahgar commented on torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:
```diff
@@ -227,9 +263,44 @@ void insertDynamicShapesGuard(
+        push(stack, std::move(elem));
+      }
+    } else {
+      at::ArrayRef<IValue> outputs = last(stack, num_ten_inputs);
```
Why are we getting the last #inputs from the stack here? Shouldn't we getting the last #outputs instead?

eellison replied:
#inputs == #outputs. See code comment I added inline

navahgar replied:
Thanks for the comments. This makes sense now :)

navahgar commented on torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:
```diff
@@ -227,9 +263,44 @@ void insertDynamicShapesGuard(
+    } else {
+      at::ArrayRef<IValue> outputs = last(stack, num_ten_inputs);
+      for (size_t i = 0; i < inputs.size(); ++i) {
+        IValue out = outputs[i];
```
Are you expecting to have the same # of outputs as inputs? Why?

eellison replied:
See code comment

Commit:
Update on "Add copy out to the fallback path in SR invocation of composed op"


We had previously handled reusing memory in the optimized kernel execution path, but not yet handled it if we hit the unoptimized fallback.



[ghstack-poisoned]

navahgar reviewed (approved):
LGTM

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458652).

Commit:
Update on "Add copy out to the fallback path in SR invocation of composed op"


We had previously handled reusing memory in the optimized kernel execution path, but not yet handled it if we hit the unoptimized fallback.

Differential Revision: [D33458652](https://our.internmc.facebook.com/intern/diff/D33458652)

[ghstack-poisoned]

Commit:
Update on "Add copy out to the fallback path in SR invocation of composed op"


We had previously handled reusing memory in the optimized kernel execution path, but not yet handled it if we hit the unoptimized fallback.

Differential Revision: [D33458652](https://our.internmc.facebook.com/intern/diff/D33458652)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458652).

Commit:
Update on "Add copy out to the fallback path in SR invocation of composed op"


We had previously handled reusing memory in the optimized kernel execution path, but not yet handled it if we hit the unoptimized fallback.

Differential Revision: [D33458652](https://our.internmc.facebook.com/intern/diff/D33458652)

[ghstack-poisoned]

Commit:
Update on "Add copy out to the fallback path in SR invocation of composed op"


We had previously handled reusing memory in the optimized kernel execution path, but not yet handled it if we hit the unoptimized fallback.

Differential Revision: [D33458652](https://our.internmc.facebook.com/intern/diff/D33458652)

[ghstack-poisoned]

Commit:
Update on "Add copy out to the fallback path in SR invocation of composed op"


We had previously handled reusing memory in the optimized kernel execution path, but not yet handled it if we hit the unoptimized fallback.

Differential Revision: [D33458652](https://our.internmc.facebook.com/intern/diff/D33458652)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458652).
